distribution/lib/Standard/Base/0.0.0-dev/src/Runtime/Lazy.enso

```diff
@@ -0,0 +1,75 @@
+import project.Any.Any
+import project.Error.Error
+import project.Nothing.Nothing
+import project.Panic.Caught_Panic
+import project.Panic.Panic
+import project.Runtime.Ref.Ref
+
+## Holds a value that is computed on first access.
+type Lazy
+    ## PRIVATE
+    Lazy (cached_ref : Ref) (builder : Nothing -> Any)
+
+    ## PRIVATE
+    Eager (value : Any)
+
+    ## Creates a new lazy value.
+    new : Any -> Lazy
+    new ~lazy_computation =
+        builder _ = lazy_computation
+        cached_ref = Ref.new Lazy_Not_Computed_Mark
+        Lazy.Lazy cached_ref builder
+
+    ## Creates a pre-computed lazy value.
+       This can be useful if a value needs to admit the Lazy type API, but is
+       known beforehand.
+    new_eager value = Lazy.Eager value
+
+    ## Returns the stored value.
+
+       The value will be computed on first access and cached.
+    get : Any
+    get self = case self of
+        Lazy.Lazy cached_ref builder -> case cached_ref.get of
+            Lazy_Not_Computed_Mark ->
+                cached_value = Cached_Value.freeze builder
+                cached_ref.put cached_value
+                cached_value.get
+            cached_value -> cached_value.get
```

Comment:
```suggestion
            cached_ref.get.get
```

Reply:
> first time computation' ... vs. ... the 'read cached value'

Well, that's exactly what I want to address with _"lazy atom fields"_...